packages/mui-styled-engine/src/StyledEngineProvider/StyledEngineProvider.js

```diff
@@ -39,7 +27,30 @@ const createEmotionCache = (options) => {
 // It allows developers to easily override MUI styles with other styling solutions, like CSS modules.
 let cache;
 if (typeof document === 'object') {
-  cache = createEmotionCache({ key: 'css', prepend: true });
+  let insertionPoint = document.querySelector('[name="emotion-insertion-point"]');
+  if (!insertionPoint) {
+    insertionPoint = document.createElement('meta');
+    insertionPoint.setAttribute('name', 'emotion-insertion-point');
+    insertionPoint.setAttribute('content', '');
+  }
+  const head = document.querySelector('head');
+  if (head) {
+    head.prepend(insertionPoint);
+  }
+  /**
+   * This is for client-side apps only.
+   * A custom sheet is required to make the GlobalStyles API injected above the insertion point.
+   * This is because the [sheet](https://github.com/emotion-js/emotion/blob/main/packages/react/src/global.js#L94-L99) does not consume the options.
+   */
+  class MyStyleSheet extends StyleSheet {
+    insert(rule, options) {
+      if (this.key && this.key.endsWith('global')) {
```

Comment:
Because I was wrong before that the Emotion (with `prepend`) will have the same behavior if a CustomSheet is provided.

However, the behavior for the `GlobalStyles` is unpredictable if you order the styles differently:

```js
<GlobalStyles />
<CssBaseline /> // this is also using GlobalStyles
<Typography />
```

inject the styles in different order compared to:

```js
<Typography />
<GlobalStyles />
<CssBaseline /> // this is also using GlobalStyles
```

That's why we cannot rely on `prepend` (deprecated) but to switch to `insertionPoint` + custom `insert` API

Reply:
I don't follow.

Let me rephrase my question:

- Before this PR, we were always using `prepend: true`, regardless of `this.key`'s value
- Now, we're replacing `prepend: true` with `insertionPoint`. This makes perfect sense to me
- But also, we're now checking for `this.key.endsWith('global')` to decide if we should use `insertionPoint`. We weren't doing that before when using `prepend: true`. This is what I don't understand. Why do we need to check `this.key.endsWith('global')` now?

-------

Besides that, I don't understand the code examples. In both I see `CssBaseline` **after** `GlobalStyles`, and `Typography`'s placement changes. I would expect `Typography`'s placement not to have any effect on `CssBaseline` and `GlobalStyles` insertion points. Am I wrong or is the example incorrect?

Reply:
The example code is correct.

The `insertionPoint` gives the same behavior as the `prepend: true` for styles from `styled` API, so let's move it aside from the discussion. This PR focus on the GlobalStyles.

The `MyStyleSheet` is an adjustment specific for handling global styles injection order, that's why I added `this.key.endsWith('global')` because [Emotion add it](https://github.com/emotion-js/emotion/blob/main/packages/react/src/global.tsx#L90) to mark the styles as global.

The logic that generates global styles is different from `styled` in Emotion. [Here](https://github.com/emotion-js/emotion/blob/main/packages/react/src/global.tsx#L93) you will see that the stylesheet instance is created without `insertionPoint`. So, when Emotion is about to create global styles, it will goes to [this line](https://github.com/emotion-js/emotion/blob/cce67ec6b2fc94261028b4f4778aae8c3d6c5fd6/packages/sheet/src/index.ts#L99). 

This is the reason I set `this.before = insertionPoint`, so that Emotion insert the global styles to before the insertion point.
```js
this.container.insertBefore(tag, insertionPoint)
```
and when the second global styles is inserted, it is in between the first global styles and the insertion point, which is correct.

Compare between these orders
```js
<GlobalStyles />
<CssBaseline /> // this is also using GlobalStyles
<Typography />

// becomes
<style> …GlobalStyles </style>
<style> …CssBaseline </style>
<meta name="insertion-point">
<style> …Typography </style>
```

If you render Typography before, the injection order remains the same.

```js
<Typography />
<GlobalStyles />
<CssBaseline /> // this is also using GlobalStyles

// becomes
<style> …GlobalStyles </style>
<style> …CssBaseline </style>
<meta name="insertion-point">
<style> …Typography </style>
```

Reply:
That makes sense, thanks for the explanation 🙌🏼